pkg/daemon/prefetch_manager.go

```diff
@@ -0,0 +1,723 @@
+	p := &PrefetchManager{
+		nodeName:                 nodeName,
+		nodeWriter:               nodeWriter,
+		runtimeEndpoint:          runtimeEndpoint,
+		authFilePath:             authFilePath,
+		prefetchTimeout:          prefetchTimeout,
+		minStorageAvailableBytes: minStorageAvailableBytes,
+		taskManager:              taskManager{cancelFuncs: make(map[types.UID]context.CancelFunc)},
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "prefetch-manager"),
+		backoff: wait.Backoff{
+			Steps:    maxRetries,
+			Duration: retryDuration,
+			Factor:   retryFactor,
+			Cap:      retryCap,
+		},
+	}
+
+	imageSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.handleAdd,
+		UpdateFunc: p.handleUpdate,
+		DeleteFunc: p.handleDelete,
+	})
+
+	p.syncHandler = p.sync
+	p.enqueuePinnedImageSet = p.enqueue
+
+	p.imageSetLister = imageSetInformer.Lister()
+	p.imageSetSynced = imageSetInformer.Informer().HasSynced
+
+	p.nodeLister = nodeInformer.Lister()
+	p.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	return p
+}
+
+func (p *PrefetchManager) sync(key string) error {
+	klog.Infof("Syncing PinnedImageSet %q", key)
+	imageSet, err := p.imageSetLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	imageSet = imageSet.DeepCopy()
+
+	err = p.ensurePrefetch(imageSet)
+	if err != nil && errors.Is(err, ErrorNoRetry) {
+		klog.V(4).Infof("PinnedImageSet %s will not be retried: %v", key, err)
+		return nil
+	}
+	if err != nil {
+		// TODO: replace annotation with MachineConfigNodeStatus
+		annoErr := p.updatePrefetchNodeAnnotation(imageSet, err)
+		if annoErr != nil {
+			klog.Warningf("failed to update node annotations: %v", annoErr)
+		}
+		return err
+	}
+
+	annoErr := p.updatePrefetchNodeAnnotation(imageSet, err)
+	if annoErr != nil {
+		klog.Warningf("failed to update node annotations: %v", annoErr)
+	}
+
+	return nil
+}
+
+func (p *PrefetchManager) ensurePrefetch(imageSet *mcfgv1.PinnedImageSet) error {
+	prefetchImages := imageSet.Spec.PinnedImages
+	if len(prefetchImages) == 0 {
+		return nil
+	}
+
+	err := p.ensurePinnedImageSetForNode(imageSet)
+	if err != nil {
+		klog.V(4).Infof("PinnedImageSet %s is not targeting this node: %v", imageSet.Name, err)
+		return err
+	}
+
+	// cancel any existing tasks for this PinnedImageSet
+	p.taskManager.cancel(imageSet)
+
+	// TODO: uncomment when OCPNODE-1986 is resolved
+	// err = p.ensureAllocatableStorage(context.TODO(), prefetchImages)
+	// if err != nil {
+	// 	klog.Infof("Insufficient storage available for prefetching images: %v", err)
+	// 	return err
+	// }
+
+	ctx, cancel := context.WithTimeout(context.Background(), p.prefetchTimeout)
+	p.taskManager.add(imageSet, cancel)
+	defer p.taskManager.cancel(imageSet)
+
+	err = p.startWorkerPool(ctx, prefetchImages)
```

Comment:
So, if I understand this correctly, the daemons are reacting to the pinnedimageset objects directly, and self-determining whether they should be pulling an image. Thus there are two processes happening in parallel:

1. the MCC rendering the new machineconfig and the MCD main node sync reacting to that
2. the MCD reading newly added pinnedimagesets and pulling images

Is there a strict dependency on that ordering? Is there any existing guard (sorry if I missed it) for that? And is there any additional pinnedimageset correctness needed to be processed by the controller before the daemon starts?

I guess the thought experiment is a large cluster with hundreds of nodes. While you only reload the crio daemon, each node is still sequentially processing the update with some built it delay of the machineconfig to enable pinnedimagesets (the crio toml file). This can take hours on large enough clusters, but I assume each daemon process running this would start the image pull already and potentially finish by the time the crio config updates.